### PR TITLE
fix(integrations): ArnioCleaner.fit() stores column dtypes; transform() emits UserWarning on dtype drift

### LINKED ISSUE #1278: Bug: ArnioCleaner.fit() does not store column dtypes; transform() silently accepts dtype-changed DataFrames
### What happened?

`ArnioCleaner.fit()` stores only column names and count. `transform()` validates only that the column names match. If the dtype of a column changes between fit and transform (e.g. an integer column becomes a string column after a CSV round-trip), no warning or error is raised. The pipeline proceeds with incorrect type inference in the C++ layer and either produces a wrong result or raises a cryptic low-level `TypeCastError` with no reference to the dtype mismatch.

### Area

pandas conversion

### Minimal reproduction

import pandas as pd
from arnio.integrations.sklearn import ArnioCleaner

train_df = pd.DataFrame({"age": [25, 30, 35], "score": [1.0, 2.0, 3.0]})

# Simulate dtype drift: age is now strings after a CSV round-trip
test_df = pd.DataFrame({"age": ["25", "30", "?"], "score": [1.0, 2.0, 3.0]})

cleaner = ArnioCleaner(steps=[("cast_types", {"age": "int64"})])
cleaner.fit(train_df)

# Expected: UserWarning about dtype change on column 'age'
# Actual: no warning, then cryptic TypeCastError from C++
result = cleaner.transform(test_df)

### Expected behavior

`transform()` should emit a `UserWarning` when a column's dtype differs from the dtype observed during `fit()`, clearly naming the affected column and describing the change. This matches the contract of `sklearn.preprocessing.StandardScaler` and other sklearn transformers that validate input consistency in `transform()`.

### Actual behavior

No warning is raised. The pipeline silently proceeds and either produces an incorrect result or raises a low-level `TypeCastError` from inside the C++ layer with no mention of the sklearn transformer or the dtype mismatch.

### Impact

Blocks installation or import

### Arnio version

1.18.0

### Install method

pip

### Operating System

Other

### Python version

3.9+

### Environment details

```shell
Suggested fix: add `self.feature_dtypes_in_ = {col: str(X[col].dtype) for col in X.columns}` in `fit()` and compare in `transform()`, emitting `warnings.warn` on mismatch.

Relevant file: `arnio/integrations/sklearn.py` — `fit()` line 38, `transform()` line 44.
```

### Pre-flight checks

- [x] I searched existing issues and did not find a duplicate.
- [x] I included a minimal reproduction or explained why one is not possible.

### Additional context

Suggested fix: add `self.feature_dtypes_in_ = {col: str(X[col].dtype) for col in X.columns}` in `fit()` and compare in `transform()`, emitting `warnings.warn` on mismatch.

Relevant file: `arnio/integrations/sklearn.py` — `fit()` line 38, `transform()` line 44.

I am a GSSoC 2026 contributor — @MehtabSandhu11. I would like to be assigned this issue.

### PR DESCRIPTION
**Pull request by MehtabSandhu11**
## Summary
`ArnioCleaner.fit()` stored only column names and count. If a column's dtype changed between fit and transform (e.g. an integer column became strings after a CSV round-trip), `transform()` raised no warning and the pipeline silently produced wrong results or a cryptic C++ `TypeCastError` with no reference to the mismatch.

This PR adds `self.feature_dtypes_in_` in `fit()` and a per-column dtype comparison in `transform()` that emits a `UserWarning` naming the column and showing the old → new dtype. Warning only — no hard errors — as requested by the maintainer.

## Linked issue
Fixes #1278

## Type of change
- [x] Bug fix
- [ ] Feature
- [ ] Performance improvement
- [ ] Documentation
- [ ] Tests
- [ ] Refactor
- [ ] CI / packaging

## Area
- [ ] CSV parser / I/O
- [ ] C++ cleaning engine
- [ ] Python API / ArFrame
- [ ] Pipeline / custom steps
- [ ] Data quality / profiling
- [ ] Schema validation
- [x] pandas interoperability
- [ ] Docs / website
- [ ] Developer tooling

## Testing
- [x] `make test`
- [x] `make lint`
- [ ] Other:

Six new tests appended to `tests/test_sklearn.py` covering:
- `transform()` emits `UserWarning` when a column dtype changes since `fit()`
- No warning is emitted when dtypes are unchanged
- The warning message names the specific column that changed
- `transform()` still returns a valid result despite the warning
- `feature_dtypes_in_` is populated correctly for all columns after `fit()`
- One warning is emitted per changed column when multiple columns change dtype

## Screenshots / Media
N/A — no UI or terminal output changes.

## Performance Impact
Negligible. `fit()` now builds a small dict of `{column_name: dtype_string}` pairs. `transform()` does one string comparison per column before the pipeline runs. Both are O(columns) and take microseconds even on wide DataFrames.

## Contributor checklist
- [x] I read the contributing guide.
- [x] I kept the PR focused on one issue or one logical change.
- [x] I added or updated tests for behavior changes.
- [x] I added or updated docs/examples for public API changes.
- [x] I checked that generated files, logs, and local build artifacts are not committed.
- [x] My PR title uses Conventional Commits (`feat:`, `fix:`, `docs:`, `test:`, `refactor:`, `chore:`).

## Maintainer notes
- Warning only, no hard errors — aligns exactly with the maintainer's instruction in the issue: "this should stay a warning-only consistency signal".
- One `UserWarning` is emitted per changed column so callers can filter on column name if needed.
- `feature_dtypes_in_` uses `str(dtype)` so it works uniformly across numpy dtypes, pandas extension types, and object columns without special-casing.
- No changes to pipeline state semantics, `allow_row_count_change`, or any existing test expectations.

### CODE REVIEW

**Review by im-anishraj — CHANGES_REQUESTED** [review]
Thanks for the PR. The implementation direction matches #1278 and `python -m py_compile arnio/integrations/sklearn.py` plus `ruff check arnio/integrations/sklearn.py tests/test_sklearn.py` pass locally.

One blocker before approval: `black --check arnio/integrations/sklearn.py tests/test_sklearn.py` would reformat both touched files. Please run Black on those files and push the formatting update; then I can re-review the behavior and tests.

**Comment by im-anishraj** [comment]
Thanks for the update. The Black formatting blocker from my earlier review looks addressed in the diff.

I approved the pending workflow runs for the current head so CI can run. I’ll wait for the visible checks to finish green before approval.

**Review by im-anishraj — APPROVED** [review]
Thanks for the formatting follow-up. I re-reviewed the final diff and ran the focused local gate on a clean PR worktree:
- `py_compile arnio/integrations/sklearn.py`: passed
- `black --check arnio/integrations/sklearn.py tests/test_sklearn.py`: passed
- `ruff check arnio/integrations/sklearn.py tests/test_sklearn.py`: passed
- `pytest tests/test_sklearn.py -v --tb=short`: 20 passed

The full GitHub CI matrix is green too. The warning-only dtype drift behavior matches the accepted #1278 scope, so approving.

**Comment by MehtabSandhu11** [comment]
@im-anishraj Please add the deserved labels along with gssoc:approved label so that I may get points in GSSoC

**Comment by im-anishraj** [comment]
Added the GSSoC approval/scoring labels to this merged PR. Thanks again for the focused contribution.
